### PR TITLE
Imviz: Change Contrast and Bias with mouse click-and-drag

### LINKED ISSUE #516: JDAT-1391: Change Contrast and Bias with mouse click-and-drag
Add viewer mode where click-and-drag changes contrast/bias.

1. Button to engage this mode (as opposed to pan/zoom mode).
2. Cut level range (left-right)
3. Center value  (up-down)

https://jira.stsci.edu/browse/JDAT-1391

### PR DESCRIPTION
**Pull request by pllim**
Fix #516 

Reference: https://github.com/glue-viz/bqplot-image-gl/blob/master/bqplot_image_gl/interacts.py

Screenshot 2021-04-21 163649

# TODO

- [x] Does the logic look about right? Or do I not need to capture a previous state?
- [x] I don't think the drag events are getting passed into the tool. ~What did I do wrong?~ Blocked by glue-viz/bqplot-image-gl#52
  - Fixed in glue-viz/bqplot-image-gl#55 and released as part of 1.3.3
- [x] When is the movement considered up/down/left/right?
- [x] What is the behavior if user drags diagonally? 
- [x] How to grab current cut levels and center values?
- [x] How to translate `delta_X` and `delta_Y` movements to changes of cut levels/center?
- [x] Do we want an option to "reset to default"? If so, how to implement (double click, special keyboard shortcut, or what)?
- [x] Do we need unit test for this? How? Naw... MVP didn't say we need a test, so...
- [x] Button need to grab layer being displayed, not `[0]`.

### CODE REVIEW

**Comment by pllim** [comment]
From @astrofrog -- Ok so three pointers:

* There is a similar tool in Qt glue, you could potentially re-use the same icon and see how it works otherwise: https://github.com/glue-viz/glue/blob/6b968b352bc5ad68b95ad5e3bb25550782a69ee8/glue/viewers/image/qt/contrast_mouse_mode.py
* There is documentation on writing custom tools (the principles and base classes are the same for Qt and Jupyter): http://docs.glueviz.org/en/stable/customizing_guide/toolbar.html
* The MatchWCSTool in jdaviz is a good example of an actual viewer tool that already exists in jdaviz: https://github.com/spacetelescope/jdaviz/blob/main/jdaviz/configs/imviz/plugins/tools.py - and importantly shows how to sign up to mouse events: https://github.com/spacetelescope/jdaviz/blob/main/jdaviz/configs/imviz/plugins/tools.py#L23 (see https://github.com/glue-viz/glue-jupyter/blob/master/glue_jupyter/bqplot/common/viewer.py#L100 for the docstring)

**Comment by astrofrog** [comment]
> Does the logic look about right? Or do I not need to capture a previous state?

I don't think you need to worry about previous state - I think the way other tools do it, e.g. DS9 is that you find the relative x and y of the cursor in the image (so as values in the range 0 to 1) and use these to then set the bias and contrast as done in:

https://github.com/glue-viz/glue/blob/6b968b352bc5ad68b95ad5e3bb25550782a69ee8/glue/viewers/image/qt/contrast_mouse_mode.py#L34-L36

The difference here will be that I don't think 'selected_layer' is a thing so you can instead get e.g. the first layer state with ``self.viewer.layers[0].state``.

This means you don't need to worry about how to determine whether the movement is vertical or horizontal or diagonal etc - you just always change bias and contrast based on where the cursor is in the image.

> I don't think the drag events are getting passed into the tool. What did I do wrong?

I'll investigate

> When is the movement considered up/down/left/right?

See above - you don't need to worry about distinguishing these

>  What is the behavior if user drags diagonally?

As above

> How to grab current cut levels and center values?

You shouldn't need these since you can set the contrast and bias directly but if you wanted them you could get them as attributes on ``self.viewer.layers[0].state`` (they are called ``v_min`` and ``v_max``).

> How to translate delta_X and delta_Y movements to changes of cut levels/center?

As in the Qt tool above as a starting point - but I don't think you want to necessarily use delta_X and delta_Y but rather just the relative x and y in the window.

>  Do we want an option to "reset to default"? If so, how to implement (double click, special keyboard shortcut, or what)?

No strong preference from me.

> Do we need unit test for this? How?

This might not be super easy to do, but I'll have a think.

**Comment by astrofrog** [comment]
Regarding the drag events, I can't quite figure this out, maybe @maartenbreddels could take a look once he is working on this again? In the mean time to keep this moving I would suggest just getting this to work for now based on click and double click - click anywhere in the image sets the contrast and bias (where clicking in the center of the image should give something close to the defaults - see the Qt tool logic for setting the values) and double click resets it? Then we could open a separate issue to try and get the dragging behavior to work DS9-style?

**Comment by pllim** [comment]
Thanks for all the comments! I don't think I'll have time to work on this further this week anyway, so would be nice to have some updates on the drag events by next week. If not, I'll implement the clicks as you suggested.

**Comment by pllim** [comment]
This is confusing, `bqplot-image-gl` and `glue-jupyter` have different event names. How do they talk to each other? How do we talk to them?

https://github.com/glue-viz/glue-jupyter/blob/536283a88a2b107f0654dda322899b868f41fa79/glue_jupyter/bqplot/common/viewer.py#L112-L122

https://github.com/glue-viz/bqplot-image-gl/blob/bd7bac12f548d20b7f2b632fb9b1cbdbc0233497/bqplot_image_gl/interacts.py#L8-L10

File: jdaviz/configs/imviz/plugins/tools.py [hunk 1]
```diff
@@ -68,3 +68,44 @@ def on_mouse_or_key_event(self, data):
...
+            # https://github.com/glue-viz/glue/blob/master/glue/viewers/image/qt/contrast_mouse_mode.py
+            with delay_callback(state, 'bias', 'contrast'):
+                state.bias = -(x * 2 - 1.5)
+                state.contrast = 10. ** (y * 2 - 1)
+
+        elif event == 'dblclick':
+            # Restore defaults that are applied on load
+            state = self.viewer.layers[0].state
+            with delay_callback(state, 'bias', 'contrast'):
+                state.bias = 0.5
```
[the remark this answers is not in the document]
Comment (pllim) [reply jdaviz/configs/imviz/plugins/tools.py:110]: Hmm. I got the numbers from printing out the values right after loading the images. But I trust you... will change it.

[the remark this answers is not in the document]
Comment (astrofrog) [reply jdaviz/configs/imviz/plugins/tools.py]: I'll double check!

[the remark this answers is not in the document]
Comment (pllim) [reply jdaviz/configs/imviz/plugins/tools.py]: It does look more like the real default after I changed it to 0 as you suggested. 🤷 

p.s. The double click isn't my fav option to reset to defaults, but it is moot when we get the drag working, I think...

**Review by astrofrog — CHANGES_REQUESTED** [review]
I just submitted https://github.com/pllim/jdaviz/pull/2 against your branch - with that the dragging works. I also changed the bias back to 0.5 as it seems that is the default in glue.

However, this breaks panning and zooming - I know why so I'll work on fixing that in glue-jupyter.

**Comment by pllim** [comment]
@astrofrog , while we wait for glue-viz/bqplot-image-gl#52 to be fixed, do you just want to separately change the default bias, as the latter is uncontroversial and unrelated to the drag bug.

**Comment by pllim** [comment]
Re: default bias -- Turns out dropping a commit is easier. The default is 0.5 again. No action on your part is needed.

**Comment by astrofrog** [comment]
No the breaking was before the bqplot-image-gl fix. All good now!

**Review by astrofrog — DISMISSED** [review]
Looks good to me!

**Review by duytnguyendtn — CHANGES_REQUESTED** [review]
Other than this one thing, I think I'd be ready to approve. When I try to drag the slider past the "end" of the screen, the window flashes. Is there a way to get it to just stop at the end if you drag too far?

https://user-images.githubusercontent.com/25206008/118845822-1adb8300-b89a-11eb-86c4-c9246fc8a172.mp4

**Comment by pllim** [comment]
Re: window flashes

@astrofrog , is the proper fix to force no-op for `on_mouse_or_key_event` if the event x and y are out of bounds? How do you define out-of-bounds here?

Good catch, @duytnguyendtn . Thanks!

File: jdaviz/configs/imviz/plugins/tools.py [hunk 2]
```diff
@@ -68,3 +69,61 @@ def on_mouse_or_key_event(self, data):
...
+        # Note that we throttle this to 200ms here as changing the contrast
+        # and bias it expensive since it forces the whole image to be redrawn
+        if event == 'dragmove':
+            if (time.time() - self._time_last) <= 0.2:
+                return
+
+            event_x = data['domain']['x']
+            event_y = data['domain']['y']
+
+            if ((event_x < self.viewer.state.x_min) or
```
Comment (pllim) [inline jdaviz/configs/imviz/plugins/tools.py:107]: @duytnguyendtn , does it behave better for you with this check in?

@astrofrog should re-review the logic here as well.

**Review by duytnguyendtn — APPROVED** [review]
Beautiful! Thanks @pllim!

**Comment by duytnguyendtn** [comment]
I won't merge yet to give @astrofrog a chance to re-review, as requested by @pllim. I'll also rerequest a review

**Review by javerbukh — APPROVED** [review]
Works great!!

**Comment by javerbukh** [comment]
@pllim This just needs a rebase then it should be good to go.

File: jdaviz/configs/imviz/plugins/tools.py [hunk 3]
```diff
@@ -84,3 +86,61 @@ def on_mouse_or_key_event(self, data):
...
+                    (event_x >= self.viewer.state.x_max) or
+                    (event_y < self.viewer.state.y_min) or
+                    (event_y >= self.viewer.state.y_max)):
+                return
+
+            x = event_x / (self.viewer.state.x_max - self.viewer.state.x_min)
+            y = event_y / (self.viewer.state.y_max - self.viewer.state.y_min)
+            state = self.viewer.layers[0].state
```
Comment (pllim) [inline jdaviz/configs/imviz/plugins/tools.py:132]: Actually, hold on... Now that we have blinking. When I blink to the second image, this tool still is locked onto the first one. We cannot hard-code `self.viewer.layers[0].state` here. @astrofrog , what is the proper way to grab the Data layer that is being displayed?

Comment (pllim) [reply jdaviz/configs/imviz/plugins/tools.py:132]: Or maybe @maartenbreddels knows? 🙏

Comment (maartenbreddels) [reply jdaviz/configs/imviz/plugins/tools.py:132]: I think it's beging displayed in order, so you could do sth like
`layer_top = [l for l in self.viewer.layers if l.visible][-1]`
Does that help?

File: setup.cfg [hunk 4]
```diff
@@ -21,7 +21,7 @@ install_requires =
     astropy>=4.2
     traitlets>=5.0.5
     bqplot>=0.12.26
-    bqplot-image-gl>=1.3.0
+    bqplot-image-gl>=1.3.3
```
Comment (javerbukh) [inline setup.cfg:24]: My PR requires 1.4.1, so to avoid a future merge conflict...
```suggestion
    bqplot-image-gl>=1.4.1

Comment (pllim) [reply setup.cfg]: Thanks!

File: jdaviz/configs/imviz/plugins/tools.py (hunk 3 above)
Comment (pllim) [reply jdaviz/configs/imviz/plugins/tools.py]: Yes, thanks!

**Comment by pllim** [comment]
Maarten's suggestion worked (thanks!), so this works now even with blinking. Given there are 2 approvals, I am merging. We can address any follow-ups as separate PR(s).

**Pull request merged**